src/workspace/file.go

```diff
@@ -0,0 +1,151 @@
+package workspace
+
+import (
+	"bytes"
+	"strings"
+
+	"github.com/VKCOM/noverify/src/git"
+)
+
+type File struct {
+	name     string
+	contents []byte
+
+	lineRanges []git.LineRange
+	lines      [][]byte
+	linesPos   []int
+
+	autoGenerated bool
+}
+
+func NewFile(name string) *File {
+	return &File{
+		name: name,
+	}
+}
+
+func NewFileWithContents(name string, contents []byte) *File {
+	lines, linesPos := parseContents(contents)
+
+	return &File{
+		name:          name,
+		contents:      contents,
+		lines:         lines,
+		linesPos:      linesPos,
+		autoGenerated: FileIsAutoGenerated(contents),
+	}
+}
+
+func parseContents(contents []byte) (lines [][]byte, linesPos []int) {
+	lines = bytes.Split(contents, []byte("\n"))
+	linesPos = make([]int, len(lines))
+	pos := 0
+	for idx, ln := range lines {
+		linesPos[idx] = pos
+		pos += len(ln) + 1
+	}
+	return lines, linesPos
+}
+
+func (f *File) Name() string {
+	return f.name
+}
+
+func (f *File) CountLines() int {
+	return len(f.lines)
+}
+
+func (f *File) Line(index int) []byte {
+	if index < 0 || index >= len(f.lines) {
+		return nil
+	}
+
+	return f.lines[index]
+}
+
+func (f *File) CountLinePosition() int {
+	return len(f.linesPos)
+}
+
+func (f *File) LinePosition(index int) int {
+	if index < 0 || index >= len(f.linesPos) {
+		return -1
+	}
+	return f.linesPos[index]
+}
+
+func (f *File) LineRanges() []git.LineRange {
+	return f.lineRanges
+}
+
+func (f *File) SetLineRanges(r []git.LineRange) {
+	f.lineRanges = r
+}
+
+func (f *File) ResetContents(newContents []byte) {
+	f.contents = newContents
+	f.lines, f.linesPos = parseContents(f.contents)
+	f.autoGenerated = FileIsAutoGenerated(f.contents)
+}
+
+func (f *File) Contents() []byte {
+	return f.contents
+}
+
+func (f *File) Part(start, end int) []byte {
```

Comment:
I'm not 100% sure that this method is needed. Most places should just slice `f.Contents` directly and 1 place should do the index checking itself. Making `File` as a mostly-data object is preferable. If you want some fancy accessing thing, you can write a helper function on the spot. Attaching a method that is only useful in 1 place makes API bigger than we need without a good reason (other places index []byte slice without any checks).